core/src/main/java/com/linecorp/armeria/server/ServerHttp1ObjectEncoder.java

```diff
@@ -62,13 +74,27 @@ public KeepAliveHandler keepAliveHandler() {
         return keepAliveHandler;
     }
 
+    WebSocketUpgradeContext webSocketUpgradeContext() {
+        return webSocketUpgradeContext;
+    }
+
     @Override
     public ChannelFuture doWriteHeaders(int id, int streamId, ResponseHeaders headers, boolean endStream,
                                         boolean isTrailersEmpty) {
         if (!isWritable(id)) {
             return newClosedSessionFuture();
         }
 
+        if (webSocketUpgradeContext.lastWebSocketUpgradeRequestId() == id) {
```

Comment:
What do you think of adding `ServiceRequestContext.isWebSocket()` and taking advantage of it here?
```
@Override
public ChannelFuture doWriteHeaders(ServiceRequestContext ctx, int id, int streamId, ResponseHeaders headers, boolean endStream, boolean isTrailersEmpty) {

    if (ctx.isWebSocket()) {
        ...
    }
}
```